### PR TITLE
Spec Draft

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**

### CODE REVIEW

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -558,14 +564,25 @@ def step(self) -> List[RequestOutput]:
...
             "execute_model",
             seq_group_metadata_list=seq_group_metadata_list,
             blocks_to_swap_in=scheduler_outputs.blocks_to_swap_in,
             blocks_to_swap_out=scheduler_outputs.blocks_to_swap_out,
             blocks_to_copy=scheduler_outputs.blocks_to_copy,
+            draft_tokens=draft_tokens
         )
+        
+        if self.spec_handler:
+            self.spec_handler.accept(output)
```
Comment (skrider) [inline vllm/engine/llm_engine.py:582]: I think spec_handler should be stateless (besides potentially managing its own KV cache), as I think that it is a good idea to centralize state as much as possible, so perhaps have it return n (following the original algorithm from https://arxiv.org/abs/2211.17192 ), and then SequenceMetadata is updated to invalidate the cache

File: vllm/engine/spec_dec.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        input_ids_list = [_pad_left_to_max(input_ids, max_len, PAD_TOKEN_ID) for input_ids in input_ids_list]
+        return torch.tensor(input_ids_list, dtype=torch.long, device='cuda')
+    
+    # TODO: we need to align draft and target model's sampler
+    def _sample_method(self, logits):
+        temperature = 1.0
+        return torch.softmax(logits / temperature, dim=-1)
+    
+    def set_draft_tokens(self,
+                seq_group_list: List[SequenceGroupMetadata]) -> None:
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:52]: We may want to add a comment for this method to describe the purpose of it since the method name `set_draft_tokens` is not that informative and the logic is quite complex. Also, please emphasize that you will update which field in seq_group_list, otherwise others have to read all your implementation to figure out.

File: vllm/engine/spec_dec.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        logger.info(f"# of input request: {len(seq_group_list)}")
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:53]: You may want to remove this kind of logs? I am wondering if it may affect the performance since it will print multiple times even for one request. Or there is any flag that helps to avoid overhead when it is running in production env?

File: vllm/engine/spec_dec.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                                   index: int):
+            token_prob = sample.sd_draft_probs[index]
+            assert token_id in token_prob
+            return token_prob[token_id]
+        
+        def extract_target_prob_dis(seq_grou_output: SequenceGroupOutputs,
+                                    token_id: int):
+            # TODO: implement this
+            vocab_size = 50272
+            return torch.rand(1, vocab_size, device='cuda').squeeze(0)
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:99]: I would say it might be better to define these two functions out of this method, otherwise, the definition code will be run every time you call `accept`. vllm is a performance sensitive server so we might want to be carefully as we can.

File: vllm/engine/spec_dec.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            logger.info(f"Seq draft tokens: {seq.draft_token_ids}")
+            # logger.info(f"Seq draft prob: {seq.draft_token_probs}")
+        
+    def accept(self,
+               target_outputs: List[SamplerOutput]):
+        def extract_draft_prob_dis(sample: SequenceOutputs,
+                                   token_id: int,
+                                   index: int):
+            token_prob = sample.sd_draft_probs[index]
+            assert token_id in token_prob
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:92]: I am not sure if vllm itself how to handle such assertion. Typically, when you assertion failed, it will terminate the program. You don't want to terminate your whole vllm server program because of just one request. Please double check on how vllm handle this kind of case. Be careful to decide you should assert or not.

File: vllm/engine/spec_dec.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            seq = seq_group_metadata.seq_data[seq_id]
+            for j in range(self.propose_cnt):
+                draft_token = draft_tokens[j][i].item()
+                seq.draft_token_probs.append({draft_token:draft_distributions[j][i]})
+                seq.draft_token_ids.append(draft_token)
+            logger.info(f"Seq draft tokens: {seq.draft_token_ids}")
+            # logger.info(f"Seq draft prob: {seq.draft_token_probs}")
+        
+    def accept(self,
+               target_outputs: List[SamplerOutput]):
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:87]: Same as above, please add informative comment.

File: vllm/engine/spec_dec.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            assert token_id in token_prob
+            return token_prob[token_id]
+        
+        def extract_target_prob_dis(seq_grou_output: SequenceGroupOutputs,
+                                    token_id: int):
+            # TODO: implement this
+            vocab_size = 50272
+            return torch.rand(1, vocab_size, device='cuda').squeeze(0)
+        
+        # Rejection Sampling
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:101]: Perhaps you may want to write a more informative comment? You are not only rejection but also accepting right?

File: vllm/engine/spec_dec.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        for seq_group_output in target_outputs:
+            assert len(seq_group_output.samples) == 1
+            sample = seq_group_output.samples[0]
+            
+            accept_token_ids = []
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:106]: accepted_token_ids

File: vllm/engine/spec_dec.py (hunk 2 above)
Comment (pian13131) [reply vllm/engine/spec_dec.py:52]: Most of the time, you may not assume the readers of your code are very familiar with specdec.

File: vllm/engine/spec_dec.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                    accept_token_ids.append(next_token.item())
+            
+            # all proposed tokens are accepted
+            if len(accept_token_ids) == len(sample.sd_draft_ids):
+                accept_token_ids.append(sample.output_token)
+            logger.info(f"accept tokens: {accept_token_ids}")
+            
+        self.invalidate_draft_kv()
+        self.invalidate_target_kv()
+        exit(0)
```
Comment (pian13131) [inline vllm/engine/spec_dec.py:129]: So are you returning nothing?

File: vllm/sequence.py [hunk 10]
```diff
@@ -57,6 +58,9 @@ class SequenceData:
     Attributes:
         prompt_token_ids: The token IDs of the prompt.
         output_token_ids: The token IDs of the output.
+        draft_token_ids: The token IDs proposed by the draft model, 
+                    it will be added by the SpecDecWorker when proposing draft tokens
+                    it will be cleared by the SpecDecWorker after accepting tokens
```
Comment (pian13131) [inline vllm/sequence.py:63]: Perhaps you can just refer the exact method name here.

Comment (pian13131) [reply vllm/sequence.py:63]: Then others will just goto that method to read the comments to figure out the detail if they want.

File: vllm/worker/worker.py [hunk 11]
```diff
@@ -173,7 +173,9 @@ def _prepare_inputs(
             seq_id = seq_ids[0]
 
             seq_data = seq_group_metadata.seq_data[seq_id]
-            prompt_tokens = seq_data.get_token_ids()
+            
+            # prompt_tokens = seq_data.get_token_ids()
+            prompt_tokens = seq_data.get_token_ids_with_draft()
```
Comment (pian13131) [inline vllm/worker/worker.py:178]: I am wondering what if they user decide to not use SD? You force to call `get_token_ids_with_draft()` instead of `get_token_ids()`? Most of the case, when you try to add a important feature, you should make it compatible with original config, which means you should assume that we may want to disable this feature.

File: vllm/engine/llm_engine.py [hunk 12]
```diff
@@ -65,6 +66,7 @@ def __init__(
         cache_config: CacheConfig,
         parallel_config: ParallelConfig,
         scheduler_config: SchedulerConfig,
+        spec_decoding_config: Optional[SpecDecConfig],
```
Comment (pian13131) [inline vllm/engine/llm_engine.py:69]: spec_dec_config?

File: vllm/engine/llm_engine.py [hunk 13]
```diff
@@ -121,6 +123,10 @@ def __init__(
         self.num_prompt_tokens: List[Tuple[float, int]] = []
         # List of (timestamp, num_tokens)
         self.num_generation_tokens: List[Tuple[float, int]] = []
+        
+        self.spec_worker = None
```
Comment (pian13131) [inline vllm/engine/llm_engine.py:127]: declare the SpecDecWorker class here?

File: vllm/engine/arg_utils.py [hunk 14]
```diff
@@ -171,6 +174,17 @@ def add_cli_args(
...
                             help='Method used to quantize the weights')
+        
+        # speculative decoding setting
+        parser.add_argument('--draft-model',
+                            type=str,
+                            default=None,
+                            help='name or path of the huggingface model to use as the draft model')
+        parser.add_argument('--propose-cnt',
+                            type=int,
+                            default=None,
```
Comment (pian13131) [inline vllm/engine/arg_utils.py:185]: You may want to add a default value, or you have to assert the input during initialization.

File: vllm/engine/arg_utils.py [hunk 15]
```diff
@@ -200,7 +214,17 @@ def create_engine_configs(
                                            self.max_num_seqs,
                                            model_config.max_model_len,
                                            self.max_paddings)
-        return model_config, cache_config, parallel_config, scheduler_config
+ 
+        spec_decoding_config = None
```
Comment (pian13131) [inline vllm/engine/arg_utils.py:218]: spec_dec_config? Also declare the config class.

**Review by pian13131 — CHANGES_REQUESTED** [review]
Please take a look the comments!

File: vllm/model_executor/layers/sampler.py [hunk 16]
```diff
@@ -88,12 +102,69 @@ def forward(
...
         return _build_sampler_output(sample_results, input_metadata,
-                                     prompt_logprobs, sample_logprobs)
+                                     prompt_logprobs, sample_logprobs, prompt_dis)
 
+    def spec_forward(self,
+        embedding: torch.Tensor,
+        hidden_states: torch.Tensor,
+        input_metadata: InputMetadata,
+        embedding_bias: Optional[torch.Tensor] = None,
+    ) -> SamplerOutput:
```
Comment (pian13131) [inline vllm/model_executor/layers/sampler.py:115]: Perhaps you may want to add a comment to tell what is the main difference between original forward and spec_forward

File: vllm/sequence.py [hunk 17]
```diff
@@ -88,7 +94,18 @@ def get_last_token_id(self) -> int:
         if not self.output_token_ids:
             return self.prompt_token_ids[-1]
         return self.output_token_ids[-1]
-
+    
+    def get_draft_token_ids(self) -> List[int]:
+        draft_tokens = [list(tp.keys())[0] for tp in self.draft_token_probs]
+        return draft_tokens
+    
+    def get_verify_token_ids(self) -> List[int]:
```
Comment (pian13131) [inline vllm/sequence.py:102]: get_verified_token_ids

File: vllm/core/scheduler.py [hunk 18]
```diff
@@ -298,6 +299,27 @@ def fork_seq(self, parent_seq: Sequence, child_seq: Sequence) -> None:
     def free_seq(self, seq: Sequence) -> None:
         self.block_manager.free(seq)
 
+    def free_invalid_kv(self, seq: Sequence, seq_out: SequenceOutputs):
+        # always clear draft tokens
+        seq.data.draft_token_probs = []
+        
+        invalid_token_cnt = max(len(seq.data.draft_token_probs) - len(seq_out.accepted_tokens), 0)
```
[the remark this answers is not in the document]
Comment (LiuXiaoxuanPKU) [reply vllm/core/scheduler.py:306]: yes, we will clear all the probabilities of draft tokens every time we accept tokens

File: vllm/sequence.py [hunk 19]
```diff
@@ -139,6 +156,9 @@ def _append_logical_block(self) -> None:
             block_size=self.block_size,
         )
         self.logical_token_blocks.append(block)
+        
+    def _delete_logical_block(self, block: LogicalTokenBlock) -> None:
+        self.logical_token_blocks.remove(block)
```
Comment (pian13131) [inline vllm/sequence.py:161]: any reason to create this method instead of just inplace use `self.logical_token_blocks.remove(block)`?

File: vllm/sequence.py [hunk 20]
```diff
@@ -166,6 +186,18 @@ def append_token_id(
         self.output_logprobs.append(logprobs)
         self.data.append_token_id(token_id, logprobs[token_id])
 
+    # delete n tokens from the end of the sequence
+    def delete_tailing_tokens(self, n: int) -> None:
+        while n > 0:
+            assert len(self.logical_token_blocks) > 0
+            last_block = self.logical_token_blocks[-1]
+            if last_block.num_tokens < n:
```
Comment (pian13131) [inline vllm/sequence.py:194]: if last_block.num_tokens == n, I think you also need to call self._delete_logical_block(last_block)?

File: vllm/worker/worker.py [hunk 21]
```diff
@@ -233,43 +239,68 @@ def _prepare_inputs(
...
-
-                if self.sliding_window is not None:
-                    sliding_window_blocks = (self.sliding_window //
-                                             self.block_size)
-                    block_table = block_table[-sliding_window_blocks:]
+                if draft_len > 0:
+                    kv_mqa = True
+                    prompt_lens.append(prompt_len)
+                    verify_tokens = seq_data.get_verified_token_ids()
+                    verify_len = len(verify_tokens)
```
Comment (pian13131) [inline vllm/worker/worker.py:273]: I think you should use `accept` instead of `verify` in all places since difference word used for same key word would introduce ambiguity

File: vllm/worker/worker.py [hunk 22]
```diff
@@ -233,43 +239,68 @@ def _prepare_inputs(
...
-                block_number = block_table[position // self.block_size]
-                block_offset = position % self.block_size
-                slot = block_number * self.block_size + block_offset
-                slot_mapping.append([slot])
-
-                if self.sliding_window is not None:
-                    sliding_window_blocks = (self.sliding_window //
-                                             self.block_size)
-                    block_table = block_table[-sliding_window_blocks:]
+                if draft_len > 0:
```
Comment (pian13131) [inline vllm/worker/worker.py:269]: I think here you'd better add some comments like:
"Speculative Decoding enabled case: <explain the main difference of this case compared with normal case, maybe something like it need to pass in multiple tokens that are accepted in previous run>"

**Review by pian13131 — CHANGES_REQUESTED** [review]

**Comment by truenorth8** [comment]
@LiuXiaoxuanPKU Are you considering to also port these changes to TGI (huggingface's inference project), at any point?
the current MR looks great btw

**Comment by LiuXiaoxuanPKU** [comment]
> @LiuXiaoxuanPKU Are you considering to also port these changes to TGI (huggingface's inference project), at any point? the current MR looks great btw

Thanks for the attention! The PR is still in progress and the current version mainly focuses on correctness instead of performance. There are still many performance improvement spaces. We are targeting to make the PR runnable this week. The vllm team will start reviewing late this week and next week.

We will not work on porting to TGI for now. Making it work in vllm is already complicated...
